addons/binding/org.openhab.binding.nikohomecontrol/src/main/java/org/openhab/binding/nikohomecontrol/internal/protocol/NikoHomeControlCommunication.java

```diff
@@ -0,0 +1,878 @@
+                    readMessage(nhcMessage);
+                }
+            } catch (IOException e) {
+                if (!listenerStopped) {
+                    // this is not an communication stop triggered from outside this runnable
+                    logger.warn("Niko Home Control: IO error in listener");
+                    // the IO has stopped working, so we need to close cleanly and try to restart
+                    stopCommunication(true);
+                }
+            }
+        }
+
+    };
+
+    /**
+     * Method that interprets all feedback from Niko Home Control and calls appropriate handling methods.
+     *
+     * @param nhcMessage message read from Niko Home Control.
+     */
+    private void readMessage(String nhcMessage) {
+
+        Gson gson = new Gson();
+        NHCCmdAck nhcCmdAck = null;
+        NHCSimpleCmdAck nhcSimpleCmdAck = null;
+
+        logger.debug("Niko Home Control: received json {}", nhcMessage);
+
+        try {
+            // first try with simple reply form, a hashmap in the data field
+            nhcSimpleCmdAck = new NHCSimpleCmdAck();
+            nhcSimpleCmdAck = gson.fromJson(nhcMessage, NHCSimpleCmdAck.class);
+            if (nhcSimpleCmdAck.cmd != null) {
+                if (nhcSimpleCmdAck.cmd.equals("systeminfo")) {
+                    systemInfo(nhcSimpleCmdAck);
+                } else if (nhcSimpleCmdAck.cmd.equals("startevents")) {
+                    startEvents(nhcSimpleCmdAck);
+                } else if ((nhcSimpleCmdAck.cmd.equals("executeactions"))
+                        && (Integer.valueOf(nhcSimpleCmdAck.data.get("error")).equals(0))) {
+                    logger.debug("Niko Home Control: execute action success");
+                } else {
+                    logger.debug("Niko Home Control: not acted on json {}", nhcMessage);
+                }
+            } else {
+                logger.debug("Niko Home Control: not acted on json {}", nhcMessage);
+            }
+        } catch (JsonSyntaxException je) {
+            // if that JSON cannot be converted, it should be an array of hashmaps in the data field
+            nhcCmdAck = new NHCCmdAck();
+            nhcCmdAck = gson.fromJson(nhcMessage, NHCCmdAck.class);
+            if (nhcCmdAck.cmd != null) {
+                if (nhcCmdAck.cmd.equals("listlocations")) {
+                    listLocations(nhcCmdAck);
+                } else if (nhcCmdAck.cmd.equals("listactions")) {
+                    listActions(nhcCmdAck);
+                } else if (nhcCmdAck.cmd.equals("listthermostat")) {
+                    listThermostat(nhcCmdAck);
+                } else if (nhcCmdAck.cmd.equals("listthermostatHVAC")) {
+                    listThermostatHVAC(nhcCmdAck);
+                } else if (nhcCmdAck.cmd.equals("readtariffdata")) {
+                    readTariffData(nhcCmdAck);
+                } else if (nhcCmdAck.cmd.equals("getalarms")) {
```

Comment:
I kept it, as the second block does look at cmd and event, making a switch more difficult. I did put it in a separate method.